src/main/java/com/zendesk/maxwell/replication/BinlogConnectorReplicator.java

```diff
@@ -606,6 +612,27 @@ public RowMap getRow() throws Exception {
 					} else {
 						processQueryEvent(event);
 					}
+					if (sql.startsWith(BinlogConnectorEvent.XASTART) || sql.startsWith(BinlogConnectorEvent.XABEGIN)) {
+						RowMapBuffer rowMapBufferXA = getTransactionRows(event);
+						String xid = getXid(sql);
+						rowBufferHashMapXA.put(xid, rowMapBufferXA);
+						LOGGER.info(BinlogConnectorEvent.XASTART + " rowBufferHashMapXA.size(): " + rowBufferHashMapXA.size() + " xid " + xid);
+					}
+					if (sql.startsWith(BinlogConnectorEvent.XACOMMIT)) {
+						String xid = getXid(sql);
+						rowBuffer = rowBufferHashMapXA.remove(xid);
+						updateStatus(rowBuffer);
```

Comment:
ok, full stop.  Can't there be multiple `XA COMMIT` events per `XA BEGIN / XA END` event?  If there can be, we'd be throwing away any previous XA transactions....

Reply:
Yes!We only deal with XA COMMIT event  and XA ROLLBACK event!

Here are some online system logs

> 19/09/22 18:12:00 INFO replication.BinlogConnectorReplicator: XA START rowBufferHashMapXA.size(): 1 xid 31302e32372e3231352e3230302e746d313536393134373132303130323238323436
19/09/22 18:12:00 INFO replication.BinlogConnectorReplicator: XA COMMIT rowBufferHashMapXA.size(): 0 xid 31302e32372e3231352e3230302e746d313536393134373132303130323238323436
19/09/22 18:12:00 INFO replication.BinlogConnectorReplicator: XA END
19/09/22 18:12:00 INFO replication.BinlogConnectorReplicator: XA START rowBufferHashMapXA.size(): 1 xid 31302e32372e3231352e3230302e746d313536393134373132303230323238323437
19/09/22 18:12:00 INFO replication.BinlogConnectorReplicator: XA END
19/09/22 18:12:00 INFO replication.BinlogConnectorReplicator: XA START rowBufferHashMapXA.size(): 2 xid 31302e32372e392e322e746d313536393134373132303231383233303133
19/09/22 18:12:00 INFO replication.BinlogConnectorReplicator: XA COMMIT rowBufferHashMapXA.size(): 1 xid 31302e32372e3231352e3230302e746d313536393134373132303230323238323437
19/09/22 18:12:00 INFO replication.BinlogConnectorReplicator: XA COMMIT rowBufferHashMapXA.size(): 0 xid 31302e32372e392e322e746d313536393134373132303231383233303133
19/09/22 18:12:01 INFO replication.BinlogConnectorReplicator: XA END
19/09/22 18:12:01 INFO replication.BinlogConnectorReplicator: XA START rowBufferHashMapXA.size(): 1 xid 31302e32372e392e322e746d313536393134373132313234333233303136
19/09/22 18:12:01 INFO replication.BinlogConnectorReplicator: XA COMMIT rowBufferHashMapXA.size(): 0 xid 31302e32372e392e322e746d313536393134373132313234333233303136